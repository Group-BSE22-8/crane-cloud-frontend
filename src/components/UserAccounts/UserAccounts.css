.context-menu {
  color: #000;
  background-color: #f9f9f9;
  position: absolute;
  box-shadow: 0 4px 4px 0 rgba(117, 117, 117, 0.2);
  -webkit-touch-callout: none;
  /* iOS Safari */
  -webkit-user-select: none;
  /* Safari */
  -khtml-user-select: none;
  /* Konqueror HTML */
  -moz-user-select: none;
  /* Old versions of Firefox */
  -ms-user-select: none;
  /* Internet Explorer/Edge */
  user-select: none;
  /* Non-prefixed version, currently supported by Chrome, Edge, Opera and Firefox */
}

.BetaUser {
  display: flex;
  gap: 1rem;
}

.BetaUserText {
  text-align: center;
  font-size: 18px;
}

.creditCoin {
  width: 1rem;
  height: 1rem;
}

.creditSection {
  display: flex;
  flex-direction: row;
  justify-content: center;
  gap: .4rem;
}

.DropDownLink a{
  color: inherit;
}