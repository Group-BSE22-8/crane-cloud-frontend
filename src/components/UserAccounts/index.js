import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import getUsersList from "../../redux/actions/users";
import "./UserAccounts.css";
import Header from "../Header";
import Spinner from "../Spinner";
import InformationBar from "../InformationBar";
import SideNav from "../SideNav";
import Modal from "../Modal";
import { ReactComponent as MoreIcon } from "../../assets/images/more-verticle.svg";
import PrimaryButton from "../PrimaryButton";
import addBetaUser from "../../redux/actions/addBetaUser";
import Feedback from "../Feedback";

class UserAccounts extends Component {
  constructor() {
    super();
    this.state = {
      actionsMenu: false,
      betaUserModal: false,
      selectedUser: "",
    };
    // this.handleChange = this.handleChange.bind(this);
    // this.handleSubmit = this.handleSubmit.bind(this);
    // this.showModal = this.showModal.bind(this);
    // this.closeModal = this.closeModal.bind(this);
    this.showMenu = this.showMenu.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleBetaUserSubmit = this.handleBetaUserSubmit.bind(this);
    this.closeBetaUserModal = this.closeBetaUserModal.bind(this);
    this.showBetaUserModal = this.showBetaUserModal.bind(this);
    this.renderRedirect = this.renderRedirect.bind(this);
  }
  componentDidMount() {
    const { getUsersList } = this.props;
    getUsersList();
  }

  componentDidUpdate(prevProps) {
    const { getUsersList, isAdded } = this.props;
    if (isAdded !== prevProps.isAdded) {
      getUsersList();
      this.closeBetaUserModal();
    }
  }
  handleClick = (e) => {
    if (this.state.actionsMenu) {
      // this.closeModal();
      return;
    }
    this.setState({ actionsMenu: true });
    e.stopPropagation();
    document.addEventListener("click", this.hideModal);
  };

  hideModal = () => {
    this.setState({ actionsMenu: false });
    document.removeEventListener("click", this.hideModal);
  };

  showMenu(id) {
    this.setState({
      selectedUser: id,
    });
  }

  showBetaUserModal() {
    this.setState({
      betaUserModal: true,
    });
  }

  closeBetaUserModal() {
    this.setState({
      betaUserModal: false,
    });
  }

  handleBetaUserSubmit() {
    const { selectedUser } = this.state;
    const { addBetaUser } = this.props;
    const betaUser = {
      is_beta_user: true,
      user_id: selectedUser,
    };
    addBetaUser(betaUser);
  }

  renderRedirect = () => {
    const { isAdded } = this.props;
    if (isAdded) {
      return <Redirect to={`/accounts`} noThrow />;
    }
  };

  render() {
    const { users, isFetched, isFetching, isAdding, isFailed, error, isAdded } =
      this.props;
    const clusterName = localStorage.getItem("clusterName");
    const {
      match: { params },
    } = this.props;
    const { actionsMenu, selectedUser, betaUserModal } = this.state;
    
    return (
      <div className="MainPage">
        {isAdded ? this.renderRedirect() : null}
        <div className="TopBarSection">
          <Header />
        </div>
        <div className="MainSection">
          <div className="SideBarSection">
            <SideNav clusterName={clusterName} clusterId={params.clusterID} />
          </div>
          <div className="MainContentSection">
            <div className="InformationBarSection">
              <InformationBar header="User Accounts" showBtn={false} />
            </div>
            <div className="ContentSection">
              <div
                className={
                  isFetching
                    ? "ResourcesTable LoadingResourcesTable"
                    : "ResourcesTable"
                }
              >
                <table className="UsersTable">
                  <thead>
                    <tr>
                      <th>Name</th>
                      <th>Beta User</th>
                      <th>Email</th>
                      <th>Actions</th>
                    </tr>
                  </thead>
                  {isFetching ? (
                    <tbody>
                      <tr className="TableLoading">
                        <td>
                          <div className="SpinnerWrapper">
                            <Spinner size="big" />
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  ) : (
                    <tbody>
                      {isFetched &&
                        users !== undefined &&
                        users?.map((user) => (
                          <tr key={users.indexOf(user)}>
                            <td>{user?.name}</td>
                            <td>{user.is_beta_user ? "True" : "False"}</td>
                            <td>{user?.email}</td>
                            <td
                              onClick={(e) => {
                                this.showMenu(user.id);
                                this.handleClick(e);
                              }}
                            >
                              <MoreIcon />
                              {actionsMenu && user.id === selectedUser && (
                                <div className="BelowHeader bg-light">
                                  <div className="context-menu">
                                    <div
                                      className="DropDownLink"
                                      role="presentation"
                                      onClick={this.showBetaUserModal}
                                    >
                                      Add Beta User
                                    </div>
                                    {/* <div
                                      className="DropDownLink"
                                      role="presentation"
                                      onClick={this.showDeletionModal}
                                    >
                                      Delete
                                    </div> */}
                                  </div>
                                </div>
                              )}
                            </td>
                          </tr>
                        ))}
                    </tbody>
                  )}
                </table>
                {isFetched && users.length === 0 && (
                  <div className="NoResourcesMessage">
                    <p>No Users Available</p>
                  </div>
                )}
                {!isFetching && !isFetched && (
                  <div className="NoResourcesMessage">
                    <p>
                      Oops! Something went wrong! Failed to retrieve Available
                      Users.
                    </p>
                  </div>
                )}
              </div>
            </div>
            <Modal
              showModal={betaUserModal}
              onClickAway={this.closeBetaUserModal}
            >
              <div class="modal-content">
                <div class="modal-header">
                  <h5
                    class="modal-title BetaUserText"
                    id="exampleModalLongTitle"
                  >
                    Add user as Beta user?
                  </h5>
                </div>
                <div class="modal-footer BetaUser">
                  <PrimaryButton
                    type="button"
                    className="CancelBtn"
                    label="Cancel"
                    onClick={this.closeBetaUserModal}
                  >
                    Cancel
                  </PrimaryButton>
                  <PrimaryButton
                    type="button"
                    label={isAdding ? <Spinner /> : "Confirm"}
                    onClick={this.handleBetaUserSubmit}
                  />
                </div>
                {isFailed && (
                  <Feedback
                    message={error !== "" ? error : null}
                    type={"error"}
                  />
                )}
              </div>
            </Modal>
          </div>
        </div>
      </div>
    );
  }
}

// inititate props
UserAccounts.propTypes = {
  users: PropTypes.arrayOf(PropTypes.object),
  isFetching: PropTypes.bool,
  isFetched: PropTypes.bool,
  getUsersList: PropTypes.func.isRequired,
};

// assigning defaults
UserAccounts.defaultProps = {
  users: [],
  isFetching: false,
  isFetched: false,
};

const mapStateToProps = (state) => {
  const { isFetching, users, isFetched } = state.usersListReducer;
  const { isAdded, isAdding, isFailed, error } = state.addBetaUserReducer;
  return { isFetching, users, isFetched, isAdded, isAdding, isFailed, error };
};

const mapDispatchToProps = {
  getUsersList,
  addBetaUser,
};

export default connect(mapStateToProps, mapDispatchToProps)(UserAccounts);
