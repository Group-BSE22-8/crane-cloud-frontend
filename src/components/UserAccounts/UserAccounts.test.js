/* eslint-disable no-undef */
import React from "react";
import { shallow } from "enzyme";
import UserAccounts, { mapStateToProps } from "./";

const UserAccountsProps = {
  match: { params: { projectID: "1" } },
  location: { pathname: "path" },
  getUsersList: jest.fn(),
};

describe("Testing the User Accounts Page component", () => {
  const WrapperUserAccounts = UserAccounts.WrappedComponent;
  const UserAccountsComponent = shallow(
    <WrapperUserAccounts {...UserAccountsProps} />
  );

  it("should match the snapshot for UserAccounts after adding props", () => {
    UserAccountsComponent.setProps(UserAccounts);
    expect(UserAccountsComponent).toBeDefined();
  });
  it("matchs the UserAccounts component snapshot", () => {
    expect(UserAccountsComponent).toMatchSnapshot();
  });
  it("should check the search input field", () => {
    UserAccountsComponent.find("input")
      .props()
      .onChange({
        target: {
          name: "rhod",
          value: "rhodin",
        },
      });
    expect(UserAccountsComponent.state("input")).toBeUndefined();
  });
});

describe("Testing the exported mapstate to props and dispatch for UserAccounts", () => {
  it("matches the UserAccounts mapstostate", () => {
    expect(
      mapStateToProps({
        usersListReducer: { isFetching: false, isFetched: false },
        addBetaUserReducer: {
          isAdded: false,
          isAdding: false,
          isFailed: false,
          error: "",
        },
        addUserCreditsReducer: {
          Added: false,
          Adding: false,
          Failed: false,
        },
      })
    ).toEqual({
      isAdded: false,
      isAdding: false,
      isFetching: false,
      isFailed: false,
      isFetched: false,
      Added: false,
      Adding: false,
      Failed: false,
      error: "",
    });
  });
});
