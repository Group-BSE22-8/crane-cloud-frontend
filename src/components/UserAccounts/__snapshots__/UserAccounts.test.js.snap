// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the User Accounts Page component matchs the UserAccounts component snapshot 1`] = `ShallowWrapper {}`;
